File: v12/interpreters/ts/src/parser/diagnostics.ts
import type { DiagnosticLocation, DiagnosticNote, DiagnosticSeverity } from "../typechecker/diagnostics";
import type { Node } from "./shared";
import { MapperError } from "./shared";

export type ParserDiagnostic = {
  severity: DiagnosticSeverity;
  message: string;
  code?: string;
  location?: DiagnosticLocation;
  notes?: DiagnosticNote[];
};

export class ParserDiagnosticError extends Error {
  readonly diagnostic: ParserDiagnostic;

  constructor(diagnostic: ParserDiagnostic) {
    super(diagnostic.message);
    this.name = "ParserDiagnosticError";
    this.diagnostic = diagnostic;
  }
}

export function buildSyntaxErrorDiagnostic(root: Node, origin?: string): ParserDiagnostic {
  const missing = findFirstMissingNode(root);
  const errorNode = missing ?? findFirstErrorNode(root) ?? root;
  const location = toDiagnosticLocation(errorNode, origin);
  const expected = missing ? formatExpectedKind(missing.type) : null;
  const message = expected ? `parser: syntax error: expected ${expected}` : "parser: syntax error";
  return {
    severity: "error",
    message,
    location,
  };
}

export function buildMapperErrorDiagnostic(error: MapperError, origin?: string): ParserDiagnostic {
  const location = error.node ? toDiagnosticLocation(error.node, error.origin ?? origin) : origin ? { path: origin } : undefined;
  return {
    severity: "error",
    message: error.message,
    location,
  };
}

export function buildParserDiagnostic(error: unknown, origin?: string): ParserDiagnostic {
  if (error instanceof ParserDiagnosticError) {
    return error.diagnostic;
  }
  if (error instanceof MapperError) {
    return buildMapperErrorDiagnostic(error, origin);
  }
  if (error instanceof Error) {
    return {
      severity: "error",
      message: error.message,
      location: origin ? { path: origin } : undefined,
    };
  }
  return {
    severity: "error",
    message: String(error),
    location: origin ? { path: origin } : undefined,
  };
}

function toDiagnosticLocation(node: Node | null | undefined, origin?: string): DiagnosticLocation | undefined {
  if (!node && !origin) return undefined;
  if (!node) return origin ? { path: origin } : undefined;
  const start = node.startPosition;
  const end = node.endPosition;
  return {
    path: origin,
    line: start.row + 1,
    column: start.column + 1,
    endLine: end.row + 1,
    endColumn: end.column + 1,
  };
}

function findFirstMissingNode(root: Node): Node | null {
  let best: Node | null = null;
  walkNodes(root, (node) => {
    if (!nodeIsMissing(node)) return;
    if (!best || node.startIndex < best.startIndex) {
      best = node;
    }
  });
  return best;
}

function findFirstErrorNode(root: Node): Node | null {
  let best: Node | null = null;
  walkNodes(root, (node) => {
    if (!nodeIsError(node)) return;
    if (!best || node.startIndex < best.startIndex) {
      best = node;
    }
  });
  return best;
}

function walkNodes(root: Node, visit: (node: Node) => void): void {
  visit(root);
  for (let i = 0; i < root.childCount; i += 1) {
    const child = root.child(i);
    if (child) {
      walkNodes(child, visit);
    }
  }
}

function nodeIsMissing(node: Node): boolean {
  return Boolean((node as unknown as { isMissing?: boolean }).isMissing);
}

function nodeIsError(node: Node): boolean {
  const asAny = node as unknown as { isError?: boolean };
  return node.type === "ERROR" || Boolean(asAny.isError);
}

function formatExpectedKind(kind: string): string {
  const trimmed = kind.trim();
  if (!trimmed) {
    return "token";
  }
  const isSymbol = /^[^A-Za-z0-9_]+$/.test(trimmed);
  if (trimmed.length === 1 || isSymbol) {
    return `'${trimmed}'`;
  }
  return trimmed.replace(/_/g, " ");
}

File: v12/interpreters/go/pkg/parser/diagnostics.go
package parser

import (
	"errors"
	"fmt"
	"strings"
	"unicode"

	sitter "github.com/tree-sitter/go-tree-sitter"
)

// SourceLocation captures a source span for parser diagnostics.
type SourceLocation struct {
	Line      int
	Column    int
	EndLine   int
	EndColumn int
}

// ParseError includes a message plus a best-effort source location.
type ParseError struct {
	Message  string
	Location SourceLocation
}

func (e *ParseError) Error() string {
	return e.Message
}

func wrapParseError(node *sitter.Node, err error) error {
	if err == nil {
		return nil
	}
	var parseErr *ParseError
	if errors.As(err, &parseErr) {
		return parseErr
	}
	if node == nil {
		return err
	}
	return &ParseError{
		Message:  err.Error(),
		Location: locationForNode(node),
	}
}

func syntaxError(root *sitter.Node) *ParseError {
	missing := findFirstMissingNode(root)
	errorNode := missing
	if errorNode == nil {
		errorNode = findFirstErrorNode(root)
	}
	if errorNode == nil {
		errorNode = root
	}
	location := SourceLocation{}
	if errorNode != nil {
		location = locationForNode(errorNode)
	}
	expected := ""
	if missing != nil {
		expected = formatExpectedKind(missing.Kind())
	}
	message := "parser: syntax error"
	if expected != "" {
		message = fmt.Sprintf("parser: syntax error: expected %s", expected)
	}
	return &ParseError{
		Message:  message,
		Location: location,
	}
}

func locationForNode(node *sitter.Node) SourceLocation {
	if node == nil {
		return SourceLocation{}
	}
	start := node.StartPosition()
	end := node.EndPosition()
	return SourceLocation{
		Line:      int(start.Row) + 1,
		Column:    int(start.Column) + 1,
		EndLine:   int(end.Row) + 1,
		EndColumn: int(end.Column) + 1,
	}
}

func findFirstMissingNode(root *sitter.Node) *sitter.Node {
	var best *sitter.Node
	walkNodes(root, func(node *sitter.Node) {
		if node == nil || !node.IsMissing() {
			return
		}
		if best == nil || node.StartByte() < best.StartByte() {
			best = node
		}
	})
	return best
}

func findFirstErrorNode(root *sitter.Node) *sitter.Node {
	var best *sitter.Node
	walkNodes(root, func(node *sitter.Node) {
		if node == nil || !node.IsError() {
			return
		}
		if best == nil || node.StartByte() < best.StartByte() {
			best = node
		}
	})
	return best
}

func walkNodes(root *sitter.Node, visit func(node *sitter.Node)) {
	if root == nil {
		return
	}
	visit(root)
	for i := uint(0); i < root.ChildCount(); i++ {
		child := root.Child(i)
		if child == nil {
			continue
		}
		walkNodes(child, visit)
	}
}

func formatExpectedKind(kind string) string {
	trimmed := strings.TrimSpace(kind)
	if trimmed == "" {
		return "token"
	}
	isSymbol := true
	for _, r := range trimmed {
		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '_' {
			isSymbol = false
			break
		}
	}
	if len(trimmed) == 1 || isSymbol {
		return fmt.Sprintf("'%s'", trimmed)
	}
	return strings.ReplaceAll(trimmed, "_", " ")
}
